File: mac-client/KBFlow/App/Updater.swift
import Foundation

struct UpdateInfo {
    let version: String
    let url: URL
}

class Updater {
    static let shared = Updater()
    let currentVersion = "1.1.0"
    private let apiURL = URL(string: "https://api.github.com/repos/Krsatvik1/KB/releases/latest")!

    func checkForUpdates(completion: @escaping (UpdateInfo?) -> Void) {
        var req = URLRequest(url: apiURL)
        req.setValue("FlowDesk-Mac", forHTTPHeaderField: "User-Agent")
        URLSession.shared.dataTask(with: req) { [weak self] data, _, _ in
            guard let self,
                  let data,
                  let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let tag = json["tag_name"] as? String,
                  let urlStr = json["html_url"] as? String,
                  let url = URL(string: urlStr)
            else { completion(nil); return }
            let latest = tag.trimmingCharacters(in: .init(charactersIn: "v"))
            if self.isNewer(latest, than: self.currentVersion) {
                completion(UpdateInfo(version: latest, url: url))
            } else {
                completion(nil)
            }
        }.resume()
    }

    private func isNewer(_ a: String, than b: String) -> Bool {
        let pa = a.split(separator: ".").compactMap { Int($0) }
        let pb = b.split(separator: ".").compactMap { Int($0) }
        for i in 0..<max(pa.count, pb.count) {
            let va = i < pa.count ? pa[i] : 0
            let vb = i < pb.count ? pb[i] : 0
            if va != vb { return va > vb }
        }
        return false
    }
}

File: windows-server/src/updater.py
import urllib.request
import json

GITHUB_RELEASES_URL = "https://api.github.com/repos/Krsatvik1/KB/releases/latest"

def check_for_updates(current_version: str) -> dict | None:
    """
    Polls GitHub Releases API.
    Returns {'version': '1.x.x', 'url': '...'} if newer, else None.
    """
    try:
        req = urllib.request.Request(
            GITHUB_RELEASES_URL,
            headers={"User-Agent": "FlowDesk-Server"}
        )
        with urllib.request.urlopen(req, timeout=5) as resp:
            data = json.loads(resp.read())
        latest = data.get("tag_name", "").lstrip("v")
        url = data.get("html_url", "")
        if _is_newer(latest, current_version):
            return {"version": latest, "url": url}
    except Exception as e:
        print(f"Update check failed: {e}")
    return None

def _is_newer(latest: str, current: str) -> bool:
    def parts(v):
        try:
            return [int(x) for x in v.split(".")]
        except Exception:
            return [0]
    return parts(latest) > parts(current)
